database-commons/src/main/java/io/cdap/plugin/db/batch/source/AbstractDBSource.java

```diff
@@ -394,31 +393,36 @@ private void validate(FailureCollector collector) {
       }
     }
 
-    protected void validateSchema(Schema actualSchema) {
-      validateSchema(actualSchema, getSchema());
+    protected void validateSchema(Schema actualSchema, FailureCollector collector) {
+      validateSchema(actualSchema, getSchema(), collector);
     }
 
     @VisibleForTesting
-    static void validateSchema(Schema actualSchema, Schema configSchema) {
+    static void validateSchema(Schema actualSchema, Schema configSchema, FailureCollector collector) {
       if (configSchema == null) {
-
-        throw new InvalidConfigPropertyException("Schema should not be null or empty", SCHEMA);
-      }
-      for (Schema.Field field : configSchema.getFields()) {
-        Schema.Field actualField = actualSchema.getField(field.getName());
-        if (actualField == null) {
-          throw new InvalidConfigPropertyException(String.format("Schema field '%s' is not present in actual record",
-                                                                 field.getName()), SCHEMA);
-        }
-        Schema actualFieldSchema = actualField.getSchema().isNullable() ?
-          actualField.getSchema().getNonNullable() : actualField.getSchema();
-        Schema expectedFieldSchema = field.getSchema().isNullable() ?
-          field.getSchema().getNonNullable() : field.getSchema();
-
-        if (!actualFieldSchema.equals(expectedFieldSchema)) {
-          throw new IllegalArgumentException(
-            String.format("Schema field '%s' has type '%s' but found '%s' in input record",
-                          field.getName(), expectedFieldSchema.getType(), actualFieldSchema.getType()));
+        collector.addFailure("Schema should not be null or empty.", null)
+          .withConfigProperty(SCHEMA);
+      } else {
+        for (Schema.Field field : configSchema.getFields()) {
+          Schema.Field actualField = actualSchema.getField(field.getName());
+          if (actualField == null) {
+            collector.addFailure(
+              String.format("Schema field '%s' is not present in actual record",
+                            field.getName()), null)
+              .withOutputSchemaField(field.getName());
+          } else {
```

Comment:
similarly, we can remove a layer of nesting and make it a little more readable if it is:

```
if (actualField == null) {
  ...
  continue;
}
...
```

Reply:
done